text/0121-config-service.md

```diff
@@ -0,0 +1,161 @@
+        NONE = 0;  // For non-periodic data (client sends points whenever)
+        SEC_1 = 1;
+        SEC_5 = 5;
+        SEC_10 = 10;
+        SEC_30 = 30;
+        MIN_1 = 60;
+        MIN_5 = 300;
+        MIN_10 = 600;
+        MIN_30 = 1800;
+        HR_1 = 3600;
+        HR_2 = 7200;
+        HR_4 = 14400;
+        HR_12 = 43200;
+        DAY_1 = 86400;
+        DAY_7 = 604800;
+      }
+      CollectionPeriod period = 3;
+
+      // Optional. Additional opaque metadata associated with the schedule.
+      // Interpreting metadata is implementation specific.
+      bytes metadata = 4;
+    }
+
+    // For this iteration, since we only want one Schedule that applies to all metrics,
+    // we will have a restriction that schedules must have a length of 1, and we will
+    // not check the patterns when we apply the collection period.
+    repeated Schedule schedules = 1;
+
+  }
+  MetricConfig metric_config = 2;
+
+  // Dynamic configs specific to trace, like the sampling rate of a resource.
+  message TraceConfig {
+    // TODO: unimplemented
+    // This proposal focuses on the metrics portion, but tracing is a natural
+    // candidate for something that might be useful to configure in the future,
+    // so this field will be left here for that eventuality.
+  }
+  TraceConfig trace_config = 3;
+
+  // Optional. The client is suggested to wait this long (in seconds) before
+  // pinging the configuration service again.
+  int32 suggested_wait_time_sec = 4;
+}
+```
+
+The SDK will periodically read a config from the service using GetConfig. If it fails to do so, it will just use either the default config or the most recent successfully read config. If it reads a new config, it will apply it.
+
+Export frequency from the SDK depends on Schedules. There can only be one Schedule for now, which defines the schedule for all metrics. The schedule has a CollectionPeriod, which defines how often metrics are exported.
+
+In the future, we will add per-metric configuration. Each Schedule also has inclusion_patterns and exclusion_patterns. Any metrics that match any of the inclusion_patterns and do not match any of the exclusion_patterns will be exported every CollectionPeriod (ie. every minute). A component will be added that can export metrics that match a pattern from one Schedule at that Schedule's collection period, while exporting other metrics that match the patterns of another Schedule at that other Schedule's collection period.
+
+The collector will support a new interface for a DynamicConfig service that can be used by an SDK, allowing a custom implementation of the configuration service protocol described above, to act as an optional bridge between an SDK and an arbitrary configuration service. This interface can be implemented as a shim to support accessing remote configurations from arbitrary backends. The collector is configured to expose an endpoint for requests to the DynamicConfig service, and returns results on that endpoint.
```

Comment:
Great points. Currently in our prototype, the service is implemented as an extension on the collector (like zpages, healthcheck). If enabled, the user has the option of specifying one of two (or more, potentially) backends from which the collector gets its configs:
*  file backend: a local file containing the metric and trace configs is served. This option corresponds to the usage you mentioned in which the collector itself is the configuration service. The file is actively monitored, so any updates to the file will be propagated to clients upon their next update request.
* remote backend: the collector receives config updates from a remote backend that implements the server-side of the gRPC service. This remote backend could be another collector, or it can be third-party. In this way, the remote backend fufills the bridging concept.

I like your idea of implementing the process as a pipeline. We chose an extension originally as a simple way to cleanly develop a prototype, but as the implementation grows, it may be nice to organize it into a full pipeline.

For specific details, feel free to check out a [PR introducing the extension](https://github.com/vmingchen/opentelemetry-collector-contrib/pull/1) and [another that adds the remote backend](https://github.com/vmingchen/opentelemetry-collector-contrib/pull/2). Both are made on our intern host's fork of the project, and will be moved to the upstream contrib repo when finished.

Does this approach sound viable? Let me know what you think!